plugins/android-boot/fu-android-boot-device.c

```diff
@@ -0,0 +1,372 @@
+/*
+ * Copyright (C) 2022 Dylan Van Assche <me@dylanvanassche.be>
+ *
+ * SPDX-License-Identifier: LGPL-2.1+
+ */
+
+#include "config.h"
+
+#include <fwupdplugin.h>
+
+#include <fcntl.h>
+#include <string.h>
+
+#include "fu-android-boot-device.h"
+
+#define ANDROID_BOOT_BOOT_SLOT_ARG    "androidboot.slot_suffix"
+#define ANDROID_BOOT_SERIAL_ARG	      "androidboot.serialno"
+#define ANDROID_BOOT_UNKNOWN_VERSION  "0.0.0"
+#define ANDROID_BOOT_SECTOR_SIZE      512
+
+struct _FuAndroidBootDevice {
+	FuUdevDevice parent_instance;
+	const gchar *label;
+	const gchar *boot_slot;
+	const gchar *uuid;
+	guint64 max_size;
+};
+
+G_DEFINE_TYPE(FuAndroidBootDevice, fu_android_boot_device, FU_TYPE_UDEV_DEVICE)
+
+static void
+fu_android_boot_device_to_string(FuDevice *device, guint idt, GString *str)
+{
+	FuAndroidBootDevice *self = FU_ANDROID_BOOT_DEVICE(device);
+
+	fu_string_append(str, idt, "BootSlot", self->boot_slot);
+	fu_string_append(str, idt, "Label", self->label);
+	fu_string_append(str, idt, "UUID", self->uuid);
+	fu_string_append_kx(str, idt, "MaxSize", self->max_size);
+}
+
+static gboolean
+fu_android_boot_device_probe(FuDevice *device, GError **error)
+{
+	FuAndroidBootDevice *self = FU_ANDROID_BOOT_DEVICE(device);
+	GUdevDevice *udev_device = fu_udev_device_get_dev(FU_UDEV_DEVICE(device));
+	g_autofree gchar *serial = NULL;
+	guint64 sectors = 0;
+	guint64 size = 0;
+	GHashTable *cmdline = NULL;
+
+	/* FuUdevDevice->probe */
+	if (!FU_DEVICE_CLASS(fu_android_boot_device_parent_class)->probe(device, error))
+		return FALSE;
+
+	/* get kernel cmdline */
+	cmdline = fu_kernel_get_cmdline(error);
+
+	/* set the physical ID */
+	if (!fu_udev_device_set_physical_id(FU_UDEV_DEVICE(device), "block", error))
+		return FALSE;
+
+	/* extract boot slot if available */
+	self->boot_slot = g_hash_table_lookup(cmdline, ANDROID_BOOT_BOOT_SLOT_ARG);
```

Comment:
you need to g_strdup() the return values from g_hash_table_lookup() as when the cmdline is unrefd these are no longer valid.